### ISSUE TITLE
Beyond Dents and Scratches: Logical Constraints in Unsupervised Anomaly Detection and Localization

### ISSUE THREAD

**Issue by shnakazawa** [TeamAidemy/ds-paper-summaries#50]
Bergmann, Paul, Kilian Batzner, Michael Fauser, David Sattlegger, and Carsten Steger. 2022. “Beyond Dents and Scratches: Logical Constraints in Unsupervised Anomaly Detection and Localization.” International Journal of Computer Vision 130 (4): 947–69.
https://link.springer.com/article/10.1007/s11263-022-01578-9

- 画像における「教師なし異常検出」のための手法。
- 異常検出：汚れや傷がある（構造的異常）とか、数が違う（論理的異常）とか。
- 構造・論理異常の両パターンを含んだデータセット (MVTec LOCO AD Dataset: Logical Constraints Anomaly Detection) と、それらを検出するアルゴリズムの評価指標 (Saturated Per-Region Overlap: sPRO)を作成。
- 異常検出・位置特定のための新しい手法 (Global Context Anomaly Detection: GCAD) を提案。構造的異常と論理的異常を同時に検出！

画像処理系の案件で使えるのでは？と想定  
↑ データセットにはNonCommercialライセンスが付与されている上、わざわざ繰り返して "In particular, it is not allowed to use the dataset for commercial purposes" とまで書かれているので、難しそう。  

構造異常を単体で検出するモデルであれば、本論文よりもより精度の高い手法があるようなので、そちらの使用を検討。

どうでもいいことだが、出版日が22年2月22日。

## Abstract
> The unsupervised detection and localization of anomalies in natural images is an intriguing and challenging problem. Anomalies manifest themselves in very different ways and an ideal benchmark dataset for this task should contain representative examples for all of them. We find that existing datasets are biased towards local structural anomalies such as scratches, dents, or contaminations. In particular, they lack anomalies in the form of violations of logical constraints, e.g., permissible objects occurring in invalid locations. We contribute a new dataset based on industrial inspection scenarios that evenly covers both types of anomalies. We provide pixel-precise ground truth data for each anomalous region and define a generalized evaluation metric that addresses localization ambiguities that can arise for logical anomalies. Furthermore, we propose a novel algorithm that improves over the state of the art in the joint detection of structural and logical anomalies. It consists of a local and a global network branch. The first one inspects confined regions independent of their spatial locations in the input image and is primarily responsible for the detection of entirely new local structures. The second one learns a globally consistent representation of the training data through a bottleneck that enables the detection of violations of long-range dependencies, a key characteristic of many logical anomalies. We perform extensive evaluations on our new dataset to corroborate our claims.

(DeepL翻訳)

自然画像における教師なし異常の検出と位置特定は、興味深くかつ挑戦的な問題である。異常は非常に様々な形で現れるため、このタスクのための理想的なベンチマークデータセットには、それら全ての代表的な例が含まれている必要がある。我々は、既存のデータセットが、傷、へこみ、汚染などの局所的な構造異常に偏っていることを発見した。特に、論理的制約の違反、例えば、許可されたオブジェクトが無効な場所に存在するような異常は含まれていない。我々は、産業界の検査シナリオに基づき、両方のタイプの異常を均等にカバーする新しいデータセットを提供する。我々は、各異常領域のピクセル精度のグランドトゥルースデータを提供し、論理的な異常のために発生し得る位置決定の曖昧さに対処する一般化された評価指標を定義する。さらに、構造的な異常と論理的な異常を同時に検出する、従来の技術よりも優れた新しいアルゴリズムを提案する。このアルゴリズムは、ローカルネットワークブランチとグローバルネットワークブランチから構成される。最初のネットワークは、入力画像内の空間的位置に依存しない限定された領域を検査し、全く新しい局所構造の検出を主に担当する。もう1つは、多くの論理的異常の主要な特徴である長距離依存性の違反の検出を可能にするボトルネックを通じて、学習データのグローバルに一貫した表現を学習する。我々は、我々の主張を裏付けるために、新しいデータセットで広範な評価を行った。

## コード

データセットと評価用のコード： https://www.mvtec.com/company/research/datasets/mvtec-loco 

**※ 業務利用不可 (CC BY-NC-SA 4.0 ライセンス)** 

## 解決した課題/先行研究との比較
- 画像の上の異常を検出する手法の研究が進んでいる。
- 2つの異常のパターン
	- 構造的異常：汚れがあるとか、形が違うとか。先行研究は主にこちらを検出するもの。
	- 論理的異常：1つしか無いはずのものが2つあるとか。
- 既存の教師なし異常検出方法の多くは、事前学習済みネットワークから抽出された局所特徴の分布をモデル化しており(Bergmann et al. 2020 ; Burlina et al. 2019)、検出できるのは構造的異常。
- 一方、論理的異常の検出にはオートエンコーダーや Generative Adversarial Networks が使える可能性がある (Liu et al . . 2020)が、これらは逆に構造的異常をうまく検出できない。
- **「構造的異常」と「論理的異常」の両方を同時にカバー**できる手法を本論文では提案。

## 技術・手法のポイント
- 構造的異常と論理的異常の両方をカバーするデータセットを作成。 "MVTec LOCO AD (Logical Constraints Anomaly Detection) Dataset"
	- 5つの品目。それぞれに正常例と、構造的な異常例、論理的な異常例を含む。
	- テスト画像ごとに、異常な部分をピクセル単位でアノテーション。
- 評価指標を作成。次項で記載。
- ピクセル単位での異常箇所特定のための新しい方法 Global Context Anomaly Detection (GCAD) を提案。
	- 2つのブランチ、Global branchとlocal branch、を用意。それぞれ主に論理的異常と構造的異常の検出を目的に。
[![Image from Gyazo](https://i.gyazo.com/c258774506f5745f798295ba6d15bc0b.png)](https://gyazo.com/c258774506f5745f798295ba6d15bc0b)
		- 構造的異常は局所的な異常であることが多い→局所的に限定された画像領域の異常スコアを計算するモデルということで、"local"
		- 論理的異常は入力画像内のものの欠落や追加など、比較的距離の離れた場所の関係に依存する。大局的に画像を扱う必要があるので、 "global"
		[![Image from Gyazo](https://i.gyazo.com/76ad23e355fb5acb4bf1efc9557a9ef4.png)](https://gyazo.com/76ad23e355fb5acb4bf1efc9557a9ef4)

## 評価指標

- 構造的異常と論理的異常、両方を評価するための指標 "Saturated Per-Region Overlap (sPRO)" を新しく作成。
- 元となる評価指標: Per-Region Overlap (PRO)
	- 異常領域が検出できているか否かの評価を、ピクセルや画像単位で行うのではなく、「異常領域ごと」にTrue positiveと判定できているピクセルの割合を算出し、それを画像中に含まれる異常箇所の数でノーマライズした値。
		- ピクセル単位で評価 (Segmentation AUROC) すると、異常箇所の面積に数値が影響される。
		- 画像単位で評価 (Detection AUROC) すると、異常とマークした箇所が間違っていても、正しく判別できていると評価されてしまう。
		- PROはこれらの課題をクリア。
		- 参考: [Qiita - 外観検査向けデータセットMVTec ADの評価指標を理解する](https://qiita.com/bezilla/items/55ff2faf2704267eab33#per-resion-overlap)
- では、論理的な異常の評価をどうするか？
	- そもそも、どうアノテーションどうするの問題。例えば、対象物の数が欠損している場合、どこマークするの？
		- 本データセットでは、論理的異常については関連する領域を広くマーク (**Fig. 3**)
			- PROだとマークされている領域と予測領域が100%近く重なる必要がある！
			 - この場合、一部マークできていればよいのでは？→ "saturation threshold" というものを導入。
- sPRO
	- saturation threshold: それぞれの画像の、各異常領域ごとに予め設定した「閾値」(**Appendix D**)
		- これ人手で決めてるっぽい (**Section 3.4**)。データセット増やすの大変そう。
		- この閾値よりも広い面積が「異常」としてマークされていれば、最高スコアが返るように。
	- どのように作用する指標かは、**Fig. 4**を参照。
[![Image from Gyazo](https://i.gyazo.com/5f80b7c6d8ef518f94300da72a29c1de.png)](https://gyazo.com/5f80b7c6d8ef518f94300da72a29c1de)

## 残された課題・議論

- 他の手法と比べて「構造」「論理」両方を解いた際のスコアは高く出ているといえど、それぞれの精度的にはまだ改善の余地がある。
	- 特に構造異常の検出においては、本手法がDetection AUROC (画像単位で算出したROC曲線下の面積. 「この画像に異常があるかないか？」を正しく回答できたかの指標) で0.8程度であるのに対し、先行研究のMVTecADは0.99 を出す。
- Saturation thresholdの決め方。より適切・より省エネルギーに。

## 重要な引用
- 構造的異常の先行研究
	- Bergmann, Paul, Michael Fauser, David Sattlegger, and Carsten Steger. 2020. “Uninformed Students: Student-Teacher Anomaly Detection with Discriminative Latent Embeddings.” In 2020 IEEE/CVF Conference on Computer Vision and Pattern Recognition (CVPR). IEEE. https://doi.org/10.1109/cvpr42600.2020.00424.
		- MVTecAD
	- Burlina, Philippe, Neil Joshi, and I-Jeng Wang. 2019. “Where’s Wally Now? Deep Generative and Discriminative Embeddings for Novelty Detection.” In 2019 IEEE/CVF Conference on Computer Vision and Pattern Recognition (CVPR). IEEE. https://doi.org/10.1109/cvpr.2019.01177.
- オートエンコーダーを使った異常検出の先行研究
	- Liu, Wenqian, Runze Li, Meng Zheng, Srikrishna Karanam, Ziyan Wu, Bir Bhanu, Richard J. Radke, and Octavia Camps. 2020. “Towards Visually Explaining Variational Autoencoders.” In Proceedings of the IEEE/CVF Conference on Computer Vision and Pattern Recognition, 8642–51.
- 本論文のイントロで引用されている事例も参照。
